### PR TITLE
Intial implementation for #11807

### PR DESCRIPTION
**Pull request by apollo13**
Looks good so far:
```
➜  nomad git:(main) ✗ ./bin/nomad namespace list                                         
Name     Description               Drivers
default  Default shared namespace  *
test     <none>                    docker,test
➜  nomad git:(main) ✗ ./bin/nomad job plan test.nomad                                    
Error during plan: Unexpected response code: 500 (1 error occurred:
	* used task driver 'lala' in cache[redis] is not allowed in namespace test
```

@tgross Do you mind giving it an initial quick review -- if I am on the right track I will continue with it.

Refs #11807

### CODE REVIEW

File: nomad/job_endpoint_validators.go
Comment (tgross) [inline nomad/job_endpoint_validators.go:22]: Probably. But it's a very cheap extra check so having it here lets us freely reorder hooks without worry and prevents us from accidentally crashing unit tests by passing in bad data 😁

File: nomad/structs/structs.go
Comment (tgross) [inline nomad/structs/structs.go:4964]: From our internal discussion of the RFC, I think we're going to want to have both a allowlist and a denylist. So can we add a `DisabledTaskDrivers` here as well?

File: api/namespace.go
Comment (tgross) [inline api/namespace.go:79]: Yes... and yes it's terrible and results in _way_ too much code to write for every feature. 😀  Long term we're hoping that the [OpenAPI](https://github.com/hashicorp/nomad-openapi) project we're doing will eventually be useable for internal use as well.

File: nomad/structs/structs.go
Comment (tgross) [inline nomad/structs/structs.go:4949]: Note that we'll need to add handling of this to the `Copy` and `SetHash` methods as well.

File: api/namespace.go
Comment (tgross) [inline api/namespace.go:74]: I think we just need this:

```suggestion
	Capabilities *NamespaceCapabilities `hcl:"capabilities,block"`
```

File: command/namespace_apply.go
Comment (tgross) [inline command/namespace_apply.go:48]: The RFC suggested using this because it made for a reasonable initial implementation, but after some internal discussion I think we're going to want to have a namespace spec like we do for quotas and volumes. Otherwise we end up having to go through a deprecation process on this flag later.

The [`quota apply`](https://github.com/hashicorp/nomad/blob/main/command/quota_apply.go#L83-L118) command has some decent example code of this. The spec can be the serialized version of the existing `api.Namespace` with the new `Capabilities` field.

**Review by tgross — COMMENTED** [review]
Overall this is looking really good @apollo13! I've left some comments based on our internal discussion of the RFC.

File: command/namespace_apply.go
Comment (apollo13) [reply command/namespace_apply.go:48]: How to handle the additional flags? Or rather is there is a file passed simply ignore them?

**Comment by apollo13** [comment]
@tgross is that more in line with what you were thinking? (HCL parsing is weird at best, but it seems to work :D)

**Comment by apollo13** [comment]
Btw, this could also nicely save https://github.com/hashicorp/nomad/issues/6554 -- but we'd have to modify the structs a bit.

Ie:
```
TaskDrivers = [{
  name: "docker",
  disabled: false,
  options: ["volumes", "privileged"]
}, {name: "raw_exec"} ...]
```

That said, given how dangerous most drivers can be I do not see much point in disabled drivers and would just keep an "allow list". I'd really love to fix #6554 with this as well

File: command/namespace_apply.go
Comment (tgross) [reply command/namespace_apply.go]: Maybe on the code path where we have a file we can ignore the flags but report a warning to the user that the file took precedence?

**Review by tgross — COMMENTED** [review]
Example HCL spec:

```hcl
name        = "dev"
description = "Namespace for developers"

capabilities {
  enabled_task_drivers  = ["docker", "exec"]
  disabled_task_drivers = ["raw_exec"]
}
```

Taking it for a spin:

```
$ nomad namespace apply ./namespace.hcl
Successfully applied namespace specification "dev"!

$ nomad namespace list
Name     Description               EnabledDrivers  DisabledDrivers
default  Default shared namespace  *               <none>
dev      Namespace for developers  docker,exec     raw_exec

$ nomad namespace status dev
Name            = dev
Description     = Namespace for developers
Quota           = <none>
EnabledDrivers  = docker,exec
DisabledDrivers = raw_exec
```

So far so good!

**Comment by apollo13** [comment]
@tgross So, how could we move this forward? Are there any architectural concerns? And what do you think about supporting privileged as well somehow? I'd rather think of privileged as well in this context than later determining that we have coded ourself into a corner.

**Comment by tgross** [comment]
Hi @apollo13! Sorry I haven't circled back to this, I've been swamped on another item. We had a conversation about this internally and I think we're generally happy with the idea from an architectural standpoint. I'll try to give it another detailed review shortly. Thanks for your patience (and for nudging me 😁 )

File: nomad/structs/structs.go
Comment (the-maldridge) [reply nomad/structs/structs.go]: This feels like the place where Apache's confusing `order by allow,deny allow from all` logic fits.  I think it should be you allow first, and then prune out of that set.

File: nomad/job_endpoint_validators_test.go
Comment (tgross) [inline nomad/job_endpoint_validators_test.go:15]: The code in `nomad/job_endpoint_validators.go` looks like it would be nicely served here by a "table driven" test where you pass a list of task/namespace pairs and whether they should validate or not.

**Review by tgross — CHANGES_REQUESTED** [review]
Hey @apollo13 this is looking really good! To get this out of draft status and moving forward:
* Let's write some tests.
* Would you mind adding a changelog file at `.changelog/11807.txt`? The format looks like [this](https://github.com/hashicorp/nomad/blob/main/.changelog/11889.txt).
* The changes to the command help text should get mirrored to [apply](https://github.com/hashicorp/nomad/blob/main/website/content/docs/commands/namespace/apply.mdx) and [status](https://github.com/hashicorp/nomad/blob/main/website/content/docs/commands/namespace/status.mdx) docs.

**Comment by the-maldridge** [comment]
So I've had more time to think about this in relation to #6554 and I think a really elegant solution would just be to dispense the task driver plugins multiple times.  As I understand it, you can kind of put whatever you want as the HCL block identifier for plugin configs.  For drivers that have a concept of lower privilege modes, you could just dispense the driver twice with two different configs (thinking mostly of container drivers here like docker, podman, and containerd, where the driver is aware of flags that need to be monitored in the config).  You could then only allow use of the privileged drivers in specific namespaces.

This approach would also let you do this in a mostly opaque way, since Nomad doesn't need to know which machines or drivers provide privileged access, from its perspective its just another driver.

**Comment by tgross** [comment]
> For drivers that have a concept of lower privilege modes, you could just dispense the driver twice with two different configs (thinking mostly of container drivers here like docker, podman, and containerd, where the driver is aware of flags that need to be monitored in the config). You could then only allow use of the privileged drivers in specific namespaces.

Hm, yeah I like where you're going with that. (That work would be separate from this PR though, so that we can ship this feature independently.)

**Comment by the-maldridge** [comment]
For sure would be a distinct item. I was wanting to raise it here because I think that once this is merged #6605 can be solved almost entirely by existing features and functionality.

**Comment by apollo13** [comment]
@tgross Pushed a few more changes. There is one failing test (at least) in https://app.circleci.com/pipelines/github/hashicorp/nomad/19158/workflows/fddc3c0b-1b53-467e-87d5-b6aa50d77b9d/jobs/193573 -- I am not sure how to fix it? One way to fix it would be to adjust my error message about non-existing namespace to match the one `upsertJobImpl` raises but then I am not sure if the existing test checks the wanted state…

**Comment by apollo13** [comment]
And now I killed semgrep:
```
2022-02-08T09:39:30.1063733Z === [ERROR] `/root/.local/bin/semgrep --skip-unknown-extensions --disable-nosem --json --autofix --dryrun --time --timeout-threshold 3 --config .semgrep/ --debug -o /tmp/tmpq_7xg6m2 .changelog/11807.txt nomad/job_endpoint_validators.go nomad/job_endpoint_validators_test.go api/namespace.go command/namespace_apply.go command/namespace_status.go nomad/job_endpoint.go nomad/structs/structs.go website/content/docs/commands/namespace/apply.mdx website/content/docs/commands/namespace/status.mdx` failed with exit code 7
2022-02-08T09:39:30.1065510Z 
2022-02-08T09:39:30.1066447Z This is an internal error, please file an issue at https://github.com/returntocorp/semgrep-action/issues/new/choose
```

File: nomad/job_endpoint_validators.go
Comment (tgross) [inline nomad/job_endpoint_validators.go:33]: Because a job can have multiple task groups and tasks, each with their own driver, it might be nice to collect a list disallowed task drivers and return a single error for the set of them.

File: nomad/job_endpoint_validators_test.go
Comment (tgross) [inline nomad/job_endpoint_validators_test.go:50]: Yeah, let's keep this test. It might be nice to validate that we don't have collisions when writing the namespace in `(*Namespace) Validate`, but it's probably a good idea to enforce that `disable` takes precedence on job submission too.

File: website/content/docs/commands/namespace/apply.mdx
Comment (tgross) [inline website/content/docs/commands/namespace/apply.mdx:67]:
```suggestion
$ nomad namespace apply namespace.json
```

File: nomad/job_endpoint_validators.go
Comment (tgross) [inline nomad/job_endpoint_validators.go:23]: > One way to fix it would be to adjust my error message about non-existing namespace to match the one `upsertJobImpl` raises but then I am not sure if the existing test checks the wanted state…

Arguably the test is checking the correct state but at the wrong spot, and we should have it in validation. But if we _only_ checked it during validation the namespace could be deleted out from under us while processing the RPC but before the transaction was opened for the upsert. The state store has its own tests so from the perspective of the job endpoint test it shouldn't matter where the error is bubbling up from.

The language of the error messages isn't consistent across the endpoints unfortunately. For this PR let's just use the same language as the state store error so that we can unblock this feature.

**Review by tgross — CHANGES_REQUESTED** [review]
This is looking great, @apollo13 

> And now I killed semgrep:

Weird! We're still experimenting with semgrep a bit so we won't consider that blocking this PR. ~I'll open an issue with upstream on that run.~ _Edit:_ already done by @shoenig https://github.com/returntocorp/semgrep-action/issues/505

File: nomad/job_endpoint_validators.go
Comment (apollo13) [reply nomad/job_endpoint_validators.go]: I have fixed that and would like to add a test for the validate method. Can you recommend where (I also need to configure some state since it accesses the namespace… mainly looking for examples where this is already done so I can copy & paste :D)

File: nomad/structs/structs.go
Comment (tgross) [inline nomad/structs/structs.go:5020]: Sorry, I missed these two lines were supposed to have resolved my comment here https://github.com/hashicorp/nomad/pull/11813#discussion_r782433839 but it turns out this copies the pointers the strings point to, not the strings. Which means this can happen: https://play.golang.com/p/IbahWCEniLV

We have a helper function [`CopySliceString`](https://github.com/hashicorp/nomad/blob/v1.2.6/helper/funcs.go#L383-L392) for just this use case.

File: nomad/job_endpoint_validators.go
Comment (tgross) [inline nomad/job_endpoint_validators.go:39]: The `%q` formatter quotes this nicely (and consistently with how we do it elsewhere):

```suggestion
				"used task driver %q is not allowed in namespace %s", disallowedDrivers[0], ns.Name)
```

File: nomad/job_endpoint_validators.go
Comment (tgross) [inline nomad/job_endpoint_validators.go:46]: Likewise, we could either use `%q` here around the string slice or we could use some clever `strings.Join` to reduce the number of string operations we have to do: https://play.golang.com/p/u3ZsG6x-HAa

Either way works for me.

File: nomad/job_endpoint_validators.go
Comment (tgross) [reply nomad/job_endpoint_validators.go]: A new file named `job_endpoint_validators_test.go` would be the best place for it. [`TestNamespaceEndpoint_GetNamespace`](https://github.com/hashicorp/nomad/blob/main/nomad/namespace_endpoint_test.go#L17) has a reasonable example of setting up test state.

File: nomad/structs/structs.go
Comment (apollo13) [reply nomad/structs/structs.go:5020]: Ups, I do know C so I should be better at this. Yet somehow Go is still confusing -- thank you for teaching me and will fix asap.

**Review by tgross — APPROVED** [review]
LGTM! 

I don't know why the website preview is skipping this build, but I've got our docs folks looking at it. In the meantime, I pulled the PR down locally and looked at the doc page updates and that looks good too.

Thanks for this contribution, @apollo13! This will ship in Nomad 1.3.0.

**Comment by apollo13** [comment]
Wow this is great, thank you so much!

**Pull request merged**
